pkg/cluster-runtime/installer.go

```diff
@@ -118,71 +124,126 @@ func getWorkerIPList(infraDriver infradriver.InfraDriver) []net.IP {
+		master0    = masters[0]
+		workers    = getWorkerIPList(i.infraDriver)
+		all        = append(masters, workers...)
+		image      = i.infraDriver.GetClusterImageName()
+		launchCmds = i.infraDriver.GetClusterLaunchCmds()
+	)
 
 	// distribute rootfs
 	if err := i.Distributor.DistributeRootfs(all, i.infraDriver.GetClusterRootfsPath()); err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	if err := i.runClusterHook(master0, PreInstallCluster); err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	if err := i.runHostHook(PreInitHost, all); err != nil {
-		return nil, nil, err
+		return err
 	}
 
+	extension, err := i.ImageEngine.GetSealerImageExtension(&common2.GetImageAnnoOptions{ImageNameOrID: image})
+	if err != nil {
+		return fmt.Errorf("failed to get ClusterImage: %s", err)
+	}
+
+	if extension.Type == v12.AppInstaller {
+		var cmds []string
+		if len(launchCmds) > 0 {
+			cmds = launchCmds
+		} else {
+			cmds = extension.Launch.Cmds
+		}
+
+		err = i.installKubeCluster(all)
+		if err != nil {
+			return fmt.Errorf("failed to install cluster: %s", err)
+		}
+
+		err = i.installApp(master0, cmds)
+		if err != nil {
+			return fmt.Errorf("failed to install application: %s", err)
+		}
+	} else {
+		err = i.installKubeCluster(all)
+		if err != nil {
+			return fmt.Errorf("failed to install cluster: %s", err)
+		}
+	}
+
+	if err = i.runClusterHook(master0, PostInstallCluster); err != nil {
+		return err
+	}
+
+	if err = i.runHostHook(PostInitHost, all); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *Installer) installKubeCluster(all []net.IP) error {
 	if err := i.containerRuntimeInstaller.InstallOn(all); err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	crInfo, err := i.containerRuntimeInstaller.GetInfo()
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	registryConfigurator, err := registry.NewConfigurator(i.RegistryConfig, crInfo, i.infraDriver, i.Distributor)
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
-	if err := registryConfigurator.Reconcile(all); err != nil {
-		return nil, nil, err
+	if err = registryConfigurator.Reconcile(all); err != nil {
+		return err
 	}
 
 	registryDriver, err := registryConfigurator.GetDriver()
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	kubeRuntimeInstaller, err := kubernetes.NewKubeadmRuntime(i.KubeadmConfig, i.infraDriver, crInfo, registryDriver.GetInfo())
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
-	if err := kubeRuntimeInstaller.Install(); err != nil {
-		return nil, nil, err
+	if err = kubeRuntimeInstaller.Install(); err != nil {
+		return err
 	}
 
-	runtimeDriver, err := kubeRuntimeInstaller.GetCurrentRuntimeDriver()
-	if err != nil {
-		return nil, nil, err
-	}
+	return nil
+}
 
-	if err := i.runClusterHook(master0, PostInstallCluster); err != nil {
-		return nil, nil, err
-	}
+func (i *Installer) installApp(master0 net.IP, cmds []string) error {
```

Comment:
what's the reason for the master0 IP is needed here for install application

In fact, we should allow the application image to be deployed on either node

Reply:
agree with you, application image  should  to be deployed on any master node. here i just chose master0 to execute launch cmds. Do we have some scenarios that require this execute node?